relay/cache/cached_accessor.go

```diff
@@ -114,11 +116,42 @@ func (c *cachedAccessor[K, V]) Get(key K) (V, error) {
 
 	if alreadyLoading {
 		// The result is being fetched on another goroutine. Wait for it to finish.
-		result.wg.Wait()
-		return result.value, result.err
+		return c.waitForResult(ctx, result)
 	} else {
 		// We are the first goroutine to request this key.
+		return c.fetchResult(ctx, key, result)
+	}
+}
 
+// waitForResult waits for the result of a lookup that was initiated by another requester and returns it
+// when it becomes is available. This method will return quickly if the provided context is cancelled.
+// Doing so does not disrupt the other requesters that are also waiting for this result.
+func (c *cachedAccessor[K, V]) waitForResult(ctx context.Context, result *accessResult[V]) (V, error) {
+	wgChan := make(chan struct{}, 1)
+	go func() {
```

Comment:
I think this goroutine will leak if context times out first.
If we replaced waitgroup with semaphore, I think you can do something like `sem.Acquire(ctx, N)`

Reply:
I don't think this leaks. The wait group is completed inside `fetchResult()` on line 175 when the requested data has either been fetched or has returned an error while being fetched. That goroutine isn't impeded if the client's context gets cancelled.

That being said, it's better to have code that is obviously correct at first glance, rather than code that you have to stare at a while to realize that it's correct. I've switched this over to using a semaphore.

Reply:
Right, what i meant is that this goroutine will survive after the parent method returns and meaninglessly block on wg until the wg completes

Reply:
Ah, right. The point I was making was that the wait group wouldn't block forever. But I agree that the goroutine would survive for longer than necessary. Regardless, this has now been refactored into a simpler solution using a semaphore.